omroneipApp/src/drvOmroneip.cpp

```diff
@@ -519,1517 +590,1382 @@ asynStatus drvOmronEIP::optimiseTags()
     return asynError;
   }
   return asynSuccess;
 }
 
 asynStatus drvOmronEIP::writeFloat64(asynUser *pasynUser, epicsFloat64 value)
 {
-  const char * functionName = "writeFloat64";
+  const char *functionName = "writeFloat64";
   int status = 0;
-  omronDrvUser_t *drvUser = (omronDrvUser_t *)pasynUser->drvUser;
+  omronDrvUser_t *drvUser = tagMap_.at(pasynUser->reason);
   int tagIndex = drvUser->tagIndex;
   int offset = drvUser->tagOffset;
-  double timeout = pasynUser->timeout*1000;
-  if (drvUser->dataType == "REAL")
-  {
+  std::string datatype = drvUser->dataType.first;
+  double timeout = pasynUser->timeout * 1000;
+  if (datatype == "REAL")
     status = plc_tag_set_float32(tagIndex, offset, (epicsFloat32)value);
-  }
-  else if (drvUser->dataType == "LREAL")
-  {
+  else if (datatype == "LREAL")
     status = plc_tag_set_float64(tagIndex, offset, value);
+  else
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, Invalid asyn interface for dtype: %s\n", driverName, functionName, datatype.c_str());
+    return asynError;
   }
-  if (status < 0) {
-    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Write attempt returned %s\n", driverName, functionName, plc_tag_decode_error(status)); 
+
+  if (status < 0)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, Write attempt returned %s\n", driverName, functionName, plc_tag_decode_error(status));
     return asynError;
   }
   status = plc_tag_write(tagIndex, timeout);
-  if (status < 0) {
-    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Write attempt returned %s\n", driverName, functionName, plc_tag_decode_error(status)); 
+  if (status < 0)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, Write attempt returned %s\n", driverName, functionName, plc_tag_decode_error(status));
     return asynError;
   }
   return asynSuccess;
 }
 
 asynStatus drvOmronEIP::writeOctet(asynUser *pasynUser, const char *value, size_t nChars, size_t *nActual)
 {
-  const char * functionName = "writeOctet";
+  const char *functionName = "writeOctet";
   int status = 0;
-  omronDrvUser_t *drvUser = (omronDrvUser_t *)pasynUser->drvUser;
+  omronDrvUser_t *drvUser = tagMap_.at(pasynUser->reason);
   int tagIndex = drvUser->tagIndex;
   int offset = drvUser->tagOffset;
-  double timeout = pasynUser->timeout*1000;
+  std::string datatype = drvUser->dataType.first;
+  double timeout = pasynUser->timeout * 1000;
 
-  /* This is a bit ghetto because Omron does strings a bit different to what libplctag expects*/
-  if (drvUser->dataType == "STRING")
+  /* This is a bit messy because Omron does strings a bit differently to what libplctag expects*/
+  if (datatype == "STRING")
   {
-    int string_capacity = plc_tag_get_string_capacity(tagIndex, offset);
+    int string_capacity = plc_tag_get_string_capacity(tagIndex, 0);
     char stringOut[nChars + 1] = {'\0'}; // allow space for null character
```

Comment:
Does this always work? Allocating the stringOut buffer on the stack with a dynamic variable (nChars)? Can you do that now - am I just too old to know?

Reply:
Hmm, I dont see why not? I havnt run into any issues when testing it anyway

Reply:
The stack is defined and allocated at compile time in a contiguous block. So the compiler need to work out how much memory to allocate at build-time. So generally you can't use a variable to allocate an array on the stack with `int myarray[my_dynamic_var]`  It should cause a compiler error.

... but maybe that was just in the good old days and compilers are now much smarter at working these things out. Maybe `writeOctet` is only ever called with a constant `nChars` variable?